src/UtilityFunctions.py

```diff
@@ -0,0 +1,126 @@
+
+            if (piece == Piece.W_K or piece == Piece.B_K):
+                coord_king = Coord(par_space)
+
+                if (abs(coord.c - coord_king.c) + abs(coord.r - coord_king.r) == 1):
+                    check = True
+                    return
+
+    ForEachSpaceHorizontalAndVertical(h_and_v, par_board, par_king)
+
+    if check:
+        return check
+
+    # Check Diagonal movements
+    def diag(par_space: Space):
+        nonlocal check
+        piece = par_board[par_space]
+        if (piece != Piece.___ and ToPlayer(piece) != par_player):
+            if (piece == Piece.W_B or piece == Piece.B_B or piece == Piece.W_Q or piece == Piece.B_Q):
+                check = True
+                return
+
+            coord_piece = Coord(par_space)
+            if (abs(coord.c - coord_piece.c) + abs(coord.r - coord_piece.r) == 2):
+                if (piece == Piece.W_K or piece == Piece.B_K):
+                    check = True
+                    return
+                if (piece == Piece.W_P or piece == Piece.B_P):
+                    if ((par_player == Player.White and (coord_piece.r == coord.r + 1)) or (par_player == Player.Black and (coord_piece.r == coord.r - 1))):
+                        check = True
+                        return
+
+    ForEachSpaceDiagonal(diag, par_board, par_king)
+
+    if check:
+        return check
+
+    # Check L movements
+    def L(par_space: Space):
+        nonlocal check
+        piece = par_board[par_space]
+        if (piece != Piece.___ and ToPlayer(piece) != par_player):
+            if (piece == Piece.W_N or piece == Piece.B_N):
+                check = True
+                return
+
+    ForEachSpaceL(L, par_king)
+    return check
+
+
+def ForEachSpaceHorizontalAndVertical(par_function, par_board: numpy.array, par_space: Space):
+    coord = Coord(par_space)
+
+    for h in range(-1, 2):
+        for v in range(-1, 2):
+            if abs(h) + abs(v) != 1:
+                continue
+
+            dest = Coord(int(coord.c + h), int(coord.r + v))
+            while 0 <= dest.c and dest.c < 8 and 0 <= dest.r and dest.r < 8:
+                space = dest()
+                par_function(space)
+
+                if (par_board[space] != Piece.___):
+                    break
+
+                dest.c += h
+                dest.r += v
+
+
+def ForEachSpaceDiagonal(par_function, par_board: numpy.array, par_space: Space):
+    coord = Coord(par_space)
+
+    for h in range(-1, 2, 2):
+        for v in range(-1, 2, 2):
+            dest = Coord(int(coord.c + h), int(coord.r + v))
+            while 0 <= dest.c and dest.c < 8 and 0 <= dest.r and dest.r < 8:
+                space = dest()
+                par_function(space)
+
+                if (par_board[space] != Piece.___):
+                    break
+
+                dest.c += h
+                dest.r += v
+
+
+def ForEachSpaceL(par_function, par_space: Space):
+    coord = Coord(par_space)
+
+    for c in range(-2, 3, c):
```

Comment:
I think c as the 3rd parameter of range is incorrect. We want to increment by 1 each loop, so we can just specify 2 params and let the 3rd default to 1.

Reply:
Good point, misread the chess.cpp implementation